### PR TITLE
Raspawar/add tool calling

### PR DESCRIPTION
**Pull request by raspawar**
# NVIDIA Agent Tool calling

```
from llama_index.agent.nvidia import NVIDIAAgent
from llama_index.llms.nvidia import NVIDIA
from llama_index.core.tools import FunctionTool

def multiply(a: int, b: int) -> int:
    """Multiple two integers and returns the result integer"""
    return a * b

def add(a: int, b: int) -> int:
    """Add two integers and returns the result integer"""
    return a + b

multiply_tool = FunctionTool.from_defaults(fn=multiply)
add_tool = FunctionTool.from_defaults(fn=add)

llm = NVIDIA('meta/llama-3.1-70b-instruct')

agent = NVIDIAAgent.from_tools(
    [multiply_tool, add_tool], llm = llm, verbose=True
)

```

### Invoke
```
response = agent.chat("What is (121 * 3) + 42?")
print(str(response))

# inspect sources
print(response.sources)
```

Streaming
```
response = agent.stream_chat(
    "What is 121 * 2? Once you have the answer, use that number to write a"
    " story about a group of mice."
)

response_gen = response.response_gen

for token in response_gen:
    print(token, end="")
```

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [X] Yes
- [ ] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [ ] Yes
- [X] No

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [X] Added new unit/integration tests
- [X] Added new notebook (that tests end-to-end)
- [X] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I ran `make format; make lint` to appease the lint gods

cc: @mattf @dglogo @sumitkbh

### CODE REVIEW

**Review by logan-markewich — COMMENTED** [review]
Thanks for the contribution @raspawar 

High level though, i think instead of creating an entirely new agent class, we can implement the FunctionCallingLLM class on the nvidia llm class. Then, we can use the existing function calling agent, and any other modules that require function calling

**Comment by logan-markewich** [comment]
In fact, since the nvidia llm already extends openai like, it will work out of the box!

is_function_calling_model just needs to be set when the model supports function calling

**Comment by raspawar** [comment]
> Thanks for the contribution @raspawar
> 
> High level though, i think instead of creating an entirely new agent class, we can implement the FunctionCallingLLM class on the nvidia llm class. Then, we can use the existing function calling agent, and any other modules that require function calling

We would like to Leverage nvidia models to perform tasks that would be difficult to implement, such as conversational AI, query_engine, parallel function calling etc. With this agent impl it becomes quite simpler. 
I will make sure to add more examples in notebooks

**Comment by logan-markewich** [comment]
@raspawar hmm, but I'm saying, this is already possible?

For example

```
from llama_index.llms.nvidia import NVIDIA

llm = NVIDIA(..)

# as an agent
from llama_index.core.agent import FunctionCallingAgentWorker

agent = FunctionCallingAgentWorker.from_tools(tools, llm=llm)
response = agent.chat("...")

# use lower-level API
response = llm. predict_and_call(tools, user_msg="Hello!", allow_parallel_tool_calls=True)

# structured outputs w/ function calling
class MyOutput(BaseModel):
  """My output class."""
  ....

sllm = llm.as_structured_llm(MyOutput)
response = sllm.chat([ChatMessage(role="user", content="Give my MyOutput")])
print(response.message.content)
```

**Comment by raspawar** [comment]
Hey @logan-markewich, thanks for suggesting this. FunctionCallingAgentWorker does not work with NVIDIA llm class however I can use OpenAIAgentWorker in NVIDIAAgent. Would like to retain the NVIDIAAgent as we can allow only nvidia llm and add logic to support more agentic endpoints

File: llama-index-integrations/agent/llama-index-agent-nvidia/llama_index/agent/nvidia/base.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        Similar to `from_defaults` in other classes, this method will
+        infer defaults for a variety of parameters, including the LLM,
+        if they are not specified.
+
+        """
+        tools = tools or []
+
+        chat_history = chat_history or []
+        if not llm:  # default model
+            llm = NVIDIA("meta/llama-3.1-70b-instruct", is_function_calling_model=True)
```
Comment (mattf) [inline llama-index-integrations/agent/llama-index-agent-nvidia/llama_index/agent/nvidia/base.py:103]: please lift "meta/llama-3.1-70b-instruct" into a global

File: llama-index-integrations/agent/llama-index-agent-nvidia/llama_index/agent/nvidia/base.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+        if callback_manager is not None:
+            llm.callback_manager = callback_manager
+
+        memory = memory or memory_cls.from_defaults(chat_history, llm=llm)
+
+        # only llama3 endpoints support tool calling
+        # TODO: update filtering to support future models
+        if not llm.model.startswith("meta/llama-3.1-"):
```
Comment (mattf) [inline llama-index-integrations/agent/llama-index-agent-nvidia/llama_index/agent/nvidia/base.py:112]: lift this into a global table that can be checked.

model endpoints that currently support tool calling -
- nv-mistralai/mistral-nemo-12b-instruct
- meta/llama-3.1-8b-instruct
- meta/llama-3.1-70b-instruct
- meta/llama-3.1-405b-instruct
- mistralai/mistral-large-2-instruct

**Comment by logan-markewich** [comment]
@raspawar Why would `FunctionCallingAgentWorker` not work? You just need to ensure `is_function_calling_model` is set in the `LLMMetadata` when it is actually a function calling model being used

**Comment by logan-markewich** [comment]
There's nothing in here that is specific to Nvidia 😅 And if there is in the future, we can update the `_prepare_chat_with_tools()` and `get_tool_calls_from_response()` methods in the base class.

If we added an agent for every LLM, there would be 100 agent classes. Better to generalize :)

**Comment by logan-markewich** [comment]
This looks good. Just need to bump the version in the toml for the llm so that it publishes on merge 💪

**Review by logan-markewich — APPROVED** [review]

**Pull request merged**